File: scripts/ingest/fred.ts
import { z } from 'zod';
import type { MetricId, Observation, Unit } from '@dmv/shared';
import { fetchJson } from '../lib/http.js';
import { log } from '../lib/log.js';
import { IngestError } from '../lib/errors.js';
import { DMV_COUNTIES } from '../lib/counties.js';
import type { DataSource } from './DataSource.js';

const BASE_URL = 'https://api.stlouisfed.org/fred';

const FredObservationSchema = z.object({
  date: z.string(),
  value: z.string(),
});

const FredResponseSchema = z.object({
  observations: z.array(FredObservationSchema),
});

interface SeriesSpec {
  /** Either a literal series ID, or a function that builds one from county FIPS */
  seriesId: string | ((fips: string) => string);
  metric: MetricId;
  unit: Unit;
  /** "USA" → emit as a national series; "state" → use state FIPS; "county" → loop counties */
  scope: 'national' | 'state' | 'county';
}

const SERIES: readonly SeriesSpec[] = [
  // National
  { seriesId: 'MORTGAGE30US', metric: 'mortgage_30y_rate', unit: 'percent', scope: 'national' },
  { seriesId: 'MORTGAGE15US', metric: 'mortgage_15y_rate', unit: 'percent', scope: 'national' },

  // State-level FHFA HPI
  { seriesId: 'DCSTHPI', metric: 'fhfa_hpi', unit: 'index_other', scope: 'state' },
  { seriesId: 'MDSTHPI', metric: 'fhfa_hpi', unit: 'index_other', scope: 'state' },
  { seriesId: 'VASTHPI', metric: 'fhfa_hpi', unit: 'index_other', scope: 'state' },

  // County-level FHFA HPI: ATNHPIUS{FIPS}A
  {
    seriesId: (fips) => `ATNHPIUS${fips}A`,
    metric: 'fhfa_hpi',
    unit: 'index_other',
    scope: 'county',
  },

  // Realtor.com via FRED — county-level hotness score
  {
    seriesId: (fips) => `HOSCCOUNTY${fips}`,
    metric: 'hotness_score',
    unit: 'index_other',
    scope: 'county',
  },

  // Realtor.com — county median listing price
  {
    seriesId: (fips) => `MELIPRCOUNTY${fips}`,
    metric: 'median_list_price',
    unit: 'USD',
    scope: 'county',
  },
];

/** Map state-level series ID → state FIPS (for the `fips` field on observations) */
const STATE_SERIES_TO_FIPS: Record<string, string> = {
  DCSTHPI: '11',
  MDSTHPI: '24',
  VASTHPI: '51',
};

function getApiKey(): string {
  const key = process.env.FRED_API_KEY;
  if (!key) {
    throw new IngestError('FRED_API_KEY not set in environment', { source: 'fred' });
  }
  return key;
}

async function fetchSeries(seriesId: string): Promise<{ date: string; value: string }[]> {
  const apiKey = getApiKey();
  const url = `${BASE_URL}/series/observations?series_id=${seriesId}&api_key=${apiKey}&file_type=json`;

  try {
    const raw = await fetchJson(url, { label: `fred:${seriesId}` });
    const parsed = FredResponseSchema.parse(raw);
    return parsed.observations;
  } catch (err) {
    throw new IngestError(`failed to fetch FRED series ${seriesId}`, { source: 'fred', series: seriesId }, err);
  }
}

function toObservations(
  rawObs: { date: string; value: string }[],
  spec: SeriesSpec,
  fips: string,
  series: string,
): Observation[] {
  const out: Observation[] = [];
  for (const o of rawObs) {
    // FRED uses "." for missing values
    if (o.value === '.' || o.value === '') continue;
    const num = Number(o.value);
    if (!Number.isFinite(num)) continue;
    out.push({
      source: 'fred',
      series,
      fips,
      metric: spec.metric,
      observedAt: o.date,
      value: num,
      unit: spec.unit,
    });
  }
  return out;
}

export class FredSource implements DataSource {
  readonly name = 'fred';
  readonly cadence = 'monthly' as const; // mixed cadences; "monthly" represents the run cadence

  async fetch(): Promise<Observation[]> {
    const all: Observation[] = [];

    for (const spec of SERIES) {
      if (spec.scope === 'national') {
        const id = typeof spec.seriesId === 'string' ? spec.seriesId : spec.seriesId('USA');
        log.info({ series: id }, 'fetching FRED national series');
        try {
          const rawObs = await fetchSeries(id);
          all.push(...toObservations(rawObs, spec, 'USA', id));
        } catch (err) {
          log.error({ series: id, err: errMessage(err) }, 'national series failed; continuing');
        }
        continue;
      }

      if (spec.scope === 'state') {
        if (typeof spec.seriesId !== 'string') {
          log.warn({ spec }, 'state series with function ID is not supported; skipping');
          continue;
        }
        const id = spec.seriesId;
        const fips = STATE_SERIES_TO_FIPS[id];
        if (!fips) {
          log.warn({ id }, 'unknown state series; skipping');
          continue;
        }
        log.info({ series: id, fips }, 'fetching FRED state series');
        try {
          const rawObs = await fetchSeries(id);
          all.push(...toObservations(rawObs, spec, fips, id));
        } catch (err) {
          log.error({ series: id, err: errMessage(err) }, 'state series failed; continuing');
        }
        continue;
      }

      // county scope
      if (typeof spec.seriesId !== 'function') {
        log.warn({ spec }, 'county scope must use function seriesId; skipping');
        continue;
      }
      const builder = spec.seriesId;

      for (const county of DMV_COUNTIES) {
        const id = builder(county.fips);
        try {
          const rawObs = await fetchSeries(id);
          const obs = toObservations(rawObs, spec, county.fips, id);
          all.push(...obs);
          log.info({ series: id, fips: county.fips, count: obs.length }, 'fetched');
        } catch (err) {
          // Some county series don't exist for every county; warn but continue
          log.warn(
            { series: id, fips: county.fips, err: errMessage(err) },
            'county series failed; continuing',
          );
        }
        // Stay well under FRED's 120 req/min limit
        await sleep(600);
      }
    }

    return all;
  }
}

function errMessage(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: go/internal/ingest/fred/fred.go
// Package fred ports scripts/ingest/fred.ts. It fetches FRED series at three
// scopes (national, state, county) with a 600 ms inter-call sleep on county
// series to stay under FRED's 120 req/min limit. Per-series failures are
// logged and skipped, never fatal — matches TS behavior.
package fred

import (
	"context"
	"fmt"
	"log/slog"
	"net/url"
	"time"

	"github.com/meowmix1337/dmv_housing_hub/go/internal/counties"
	httpclient "github.com/meowmix1337/dmv_housing_hub/go/internal/http"
	pkglog "github.com/meowmix1337/dmv_housing_hub/go/internal/log"
	"github.com/meowmix1337/dmv_housing_hub/go/internal/types"
)

const (
	baseURL          = "https://api.stlouisfed.org/fred"
	countySleepDelay = 600 * time.Millisecond
)

type Config struct {
	APIKey string `env:"FRED_API_KEY,required"`
}

type scope int

const (
	scopeNational scope = iota
	scopeState
	scopeCounty
)

type seriesSpec struct {
	idLit  string                 // for national + state scopes
	idFn   func(fips string) string // for county scope
	metric types.MetricId
	unit   types.Unit
	scope  scope
}

var series = []seriesSpec{
	// National
	{idLit: "MORTGAGE30US", metric: types.MetricMortgage30yRate, unit: types.UnitPercent, scope: scopeNational},
	{idLit: "MORTGAGE15US", metric: types.MetricMortgage15yRate, unit: types.UnitPercent, scope: scopeNational},

	// State-level FHFA HPI
	{idLit: "DCSTHPI", metric: types.MetricFhfaHpi, unit: types.UnitIndexOther, scope: scopeState},
	{idLit: "MDSTHPI", metric: types.MetricFhfaHpi, unit: types.UnitIndexOther, scope: scopeState},
	{idLit: "VASTHPI", metric: types.MetricFhfaHpi, unit: types.UnitIndexOther, scope: scopeState},

	// County FHFA HPI: ATNHPIUS{FIPS}A
	{idFn: func(fips string) string { return "ATNHPIUS" + fips + "A" }, metric: types.MetricFhfaHpi, unit: types.UnitIndexOther, scope: scopeCounty},

	// Realtor.com via FRED — county hotness score
	{idFn: func(fips string) string { return "HOSCCOUNTY" + fips }, metric: types.MetricHotnessScore, unit: types.UnitIndexOther, scope: scopeCounty},

	// Realtor.com — county median listing price
	{idFn: func(fips string) string { return "MELIPRCOUNTY" + fips }, metric: types.MetricMedianListPrice, unit: types.UnitUSD, scope: scopeCounty},
}

var stateSeriesToFIPS = map[string]string{
	"DCSTHPI": "11",
	"MDSTHPI": "24",
	"VASTHPI": "51",
}

type fredObservation struct {
	Date  string            `json:"date"`
	Value types.MaybeFloat  `json:"value"`
}

type fredResponse struct {
	Observations []fredObservation `json:"observations"`
}

type Source struct {
	cfg        Config
	client     *httpclient.Client
	logger     *slog.Logger
	countySleep time.Duration // override for tests
	baseURL     string        // override for tests
}

func New(cfg Config, client *httpclient.Client) *Source {
	return &Source{
		cfg:         cfg,
		client:      client,
		logger:      pkglog.Default(),
		countySleep: countySleepDelay,
		baseURL:     baseURL,
	}
}

func (s *Source) Name() string            { return "fred" }
func (s *Source) Cadence() types.Cadence  { return types.CadenceMonthly }

func (s *Source) Fetch(ctx context.Context) ([]types.Observation, error) {
	var out []types.Observation
	cs := counties.All()

	for _, spec := range series {
		switch spec.scope {
		case scopeNational:
			obs, err := s.fetchSeries(ctx, spec.idLit)
			if err != nil {
				s.logger.Warn("fred national series failed", "series", spec.idLit, "err", err)
				continue
			}
			out = append(out, toObservations(obs, spec, "USA", spec.idLit)...)
			s.logger.Info("fred fetched national series", "series", spec.idLit, "count", len(obs))

		case scopeState:
			fips, ok := stateSeriesToFIPS[spec.idLit]
			if !ok {
				s.logger.Warn("fred unknown state series", "series", spec.idLit)
				continue
			}
			obs, err := s.fetchSeries(ctx, spec.idLit)
			if err != nil {
				s.logger.Warn("fred state series failed", "series", spec.idLit, "err", err)
				continue
			}
			out = append(out, toObservations(obs, spec, fips, spec.idLit)...)
			s.logger.Info("fred fetched state series", "series", spec.idLit, "fips", fips, "count", len(obs))

		case scopeCounty:
			for _, c := range cs {
				id := spec.idFn(c.FIPS)
				obs, err := s.fetchSeries(ctx, id)
				if err != nil {
					s.logger.Warn("fred county series failed", "series", id, "fips", c.FIPS, "err", err)
				} else {
					out = append(out, toObservations(obs, spec, c.FIPS, id)...)
					s.logger.Info("fred fetched county series", "series", id, "fips", c.FIPS, "count", len(obs))
				}
				if err := sleepCtx(ctx, s.countySleep); err != nil {
					return out, err
				}
			}
		}
	}

	return out, nil
}

func (s *Source) fetchSeries(ctx context.Context, seriesID string) ([]fredObservation, error) {
	q := url.Values{}
	q.Set("series_id", seriesID)
	q.Set("api_key", s.cfg.APIKey)
	q.Set("file_type", "json")
	u := s.baseURL + "/series/observations?" + q.Encode()

	var resp fredResponse
	if err := s.client.GetJSON(ctx, u, "fred:"+seriesID, &resp); err != nil {
		return nil, fmt.Errorf("fetch %s: %w", seriesID, err)
	}
	return resp.Observations, nil
}

func toObservations(in []fredObservation, spec seriesSpec, fips, seriesID string) []types.Observation {
	out := make([]types.Observation, 0, len(in))
	for _, o := range in {
		if !o.Value.Valid {
			continue // "." sentinel or null
		}
		out = append(out, types.Observation{
			Source:     "fred",
			Series:     seriesID,
			FIPS:       fips,
			Metric:     spec.metric,
			ObservedAt: o.Date,
			Value:      o.Value.Val,
			Unit:       spec.unit,
		})
	}
	return out
}

func sleepCtx(ctx context.Context, d time.Duration) error {
	if d == 0 {
		return nil
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-t.C:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

